### PR TITLE
fix(sql): FILL(LINEAR) now supports ALIGN TO CALENDAR

### LINKED ISSUE #4355: ALIGN to CALENDAR bug when using LINEAR fill
### To reproduce

when using sql that has 
`SAMPLE BY 5m  FILL(LINEAR) ALIGN TO CALENDAR`
returns result that are not aligned to CAL
ie
`2024-03-30T01:26:21.000000Z`

while it works when using FILL with NULL or PREV 

### QuestDB version:

7.4.0 , PostgreSQL 12.3, compiled by Visual C++ build 1914, 64-bit, QuestDB

### OS, in case of Docker specify Docker and the Host OS:

OSX

### File System, in case of Docker specify Host File System:

osx

### Full Name:

Bojan Tesanovic

### Affiliation:

private

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [x] #4356

### Additional context

image

### PR DESCRIPTION
**Pull request by nwoolmer**
Closes #4355 

This PR copies code used to initialise timestamp sampler from other FILL factories, into the interpolation factory. This means that it will support ALIGN TO CALENDAR, instead of always giving results for ALIGN TO FIRST OBSERVATION.

Confirmed with reporter that this gives the expected output.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 56 (73.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|41|56|73.21%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 48 / 56 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|48|56|85.71%|

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 1]
```diff
@@ -635,5 +670,76 @@ private void nullifyRange(long lo, long hi, Record record) {
...
+        }
+
+        private boolean areTimestampsInitialized;
+
+        protected void initTimestamps() {
+            if (areTimestampsInitialized) {
+                return;
+            }
+
+            if (!managedCursor.hasNext()) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:724]: is this branch ever taken? if the manager cursor has no record then there is an early return right from `buildMap()` => `initTimestamps()` is not called at all? or do I miss a possible path?

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 2]
```diff
@@ -231,12 +241,28 @@ private class SampleByInterpolateRecordCursor extends VirtualFunctionSkewedSymbo
         private long prevSample = -1;
         private long rowId;
 
+        protected final Function offsetFunc;
+        protected final int offsetFuncPos;
+        protected final TimestampSampler timestampSampler;
+        protected final Function timezoneNameFunc;
+        protected final int timezoneNameFuncPos;
+        protected long fixedOffset;
+        protected long nextDstUtc;
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:250]: it looks this field is only assigned but never read? that's strange, isn't it? also: why is it protected and not private? some other fields could be private too.

Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:250]: it seems it's entirely missing code for DST adjustment. and equivalent of this https://github.com/questdb/questdb/blob/9683d03251aa6284bf8d2de0d6d60884a477d893/core/src/main/java/io/questdb/griffin/engine/groupby/AbstractNoRecordSampleByCursor.java#L141

**Review by jerrinot — CHANGES_REQUESTED** [review]
it looks to me like there are bugs due to DST transitions being ignored. 
happy to discuss it further, just give me a shout.

**Review by jerrinot — APPROVED** [review]
Thanks to @nwoolmer for both this PR and opening https://github.com/questdb/questdb/issues/4424 to clarify the remaining things. this PR already provides value as it is.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 48 / 54 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|48|54|88.89%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 48 / 54 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|48|54|88.89%|

**Comment by javier** [comment]
Found a weird issue. Create a table using the SQL below and import the attached CSV. Then run this query

```
select timestamp, symbol,sum(price) 
from trades
where timestamp in '2024-04-22' 
sample by 1s fill(linear);
```
First thing I notice is that the timestamp order is not honored, and the first row is out of order, which does not happen with FILL(null). But weirder, if you re-execute the query, you get different results every time.


``` 
CREATE TABLE IF NOT EXISTS  'trades' (
  symbol SYMBOL capacity 256 CACHE,
  side SYMBOL capacity 256 CACHE,
  price DOUBLE,
  amount DOUBLE,
  timestamp TIMESTAMP
) timestamp (timestamp) PARTITION BY DAY WAL DEDUP UPSERT KEYS(timestamp, symbol, side);
```

[trades.csv](https://github.com/questdb/questdb/files/15062694/trades.csv)

**Pull request merged**